pvlib/horizon.py

```diff
@@ -0,0 +1,646 @@
+        A 2d array containing longitude values that correspond to the other
+        two input grids.
+
+    elev_grid : ndarray
+        A 2d array containing elevation values that correspond to the other
+        two input grids.
+
+    num_samples : tuple
+        A tuple containing two integers. The first is the desired number of
+        points along the radial axis of the polar grid. The second is the
+        desired number of points along the angular axis of the polar grid.
+
+
+    Returns
+    -------
+    all_samples: list
+       Array of [lat, lon, elev] points that were sampled using the polar grid.
+
+    """
+    assert(lat_grid.shape == lon_grid.shape == elev_grid.shape)
+
+    lats = lat_grid[0]
+    lons = lon_grid.T[0]
+
+    lat_range = lats[-1] - lats[0]
+
+    grid_shape = lat_grid.shape
+    grid_center_i = (grid_shape[0] - 1) // 2
+    grid_center_j = (grid_shape[1] - 1) // 2
+    site_lat = lat_grid[grid_center_i, grid_center_j]
+    site_lon = lon_grid[grid_center_i, grid_center_j]
+
+    interpolator = RegularGridInterpolator((lats, lons), elev_grid.T)
+
+    r = np.linspace(0, lat_range//2, num_samples[0])
+    theta = np.linspace(0, 2 * np.pi, num_samples[1])
+    polar_pts = np.array(list(itertools.product(r, theta)))
+
+    pts = np.array([tools.polar_to_cart(e[0], e[1]) for e in polar_pts])
+    pts += np.array((site_lat, site_lon))
+    total_num_samples = num_samples[0]*num_samples[1]
+
+    interpolated_elevs = interpolator(pts).reshape(total_num_samples, 1)
+    samples = np.hstack((pts, interpolated_elevs))
+    return samples
+
+
+def uniformly_sample_triangle(p1, p2, p3, num_samples):
+    """
+    Randomly sample the surface of a triangle defined by three (lat, lon, elev)
+    points uniformly [1].
+
+    Parameters
+    ----------
+    pt1 : ndarray
+        An array conaining (lat, lon, elev) values that define one vertex
+        of the triangle.
+    pt2 : ndarray
+        An array conaining (lat, lon, elev) values that define another vertex
+        of the triangle.
+    pt3 : ndarray
+        An array conaining (lat, lon, elev) values that define the last vertex
+        of the triangle.
+
+    num_samples : tuple
+        The number of random samples to be uniformly taken from the surface
+        of the triangle.
+
+    Returns
+    -------
+    points: Nx3 ndarray
+        Array with N (lat, lon, elev) points that lie on the surface of the
+        triangle.
+
+    [1] http://graphics.stanford.edu/courses/cs468-08-fall/pdf/osada.pdf
+    """
+    c1 = tools.lle_to_xyz(p1)
+    c2 = tools.lle_to_xyz(p2)
+    c3 = tools.lle_to_xyz(p3)
+
+    r1 = np.random.rand(num_samples).reshape((num_samples, 1))
+    r2 = np.random.rand(num_samples).reshape((num_samples, 1))
+    sqrt_r1 = np.sqrt(r1)
+
+    random_pts = (1-sqrt_r1)*c1 + sqrt_r1*(1-r2)*c2 + sqrt_r1*r2*c3
+    random_pts = tools.xyz_to_lle(random_pts)
+
+    return random_pts
+
+
+def filter_points(horizon_azimuths, horizon_angles, bin_size=1):
```

Comment:
"find_horizon_points" maybe?  I don't see that this function is used so I'm unsure what of it's purpose.

Reply:
The purpose of this function is to make the horizon profile smooth by consolidating horizon points that have similar azimuth values (depending on bin size). For example, if you calculate horizon points from a grid you will have many points that have an azimuth of 0. But you really only want the horizon point with the greatest elevation angle.

Reply:
That makes sense.  Maybe it would help if the inputs were `azimuths`, `elevation_angles`, and the term "horizon" is only used to for output quantities, since the inputs are not necessarily points on the visible horizon.  I do think a more specific function name would help.

Reply:
I also missed the purpose in my review, so it should be made more explicit in the documentation.